### PR TITLE
feat(core): introduce CString capacity methods

### PR DESCRIPTION
**Pull request by dkrutskikh**
### **PR Type**
Enhancement

___

### **Description**
- Add string capacity and size methods to `CString` class

- Implement `empty()`, `utf8_size()`, `length()`, `max_size()`, and `capacity()` methods

- Add comprehensive test coverage for all new methods

- Enhance `FixedString` test coverage for capacity methods

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["CString Class"] --> B["New Methods"]
  B --> C["empty()"]
  B --> D["utf8_size()"]
  B --> E["length()"]
  B --> F["max_size()"]
  B --> G["capacity()"]
  H["Test Coverage"] --> I["CString Tests"]
  H --> J["FixedString Tests"]
  I --> K["675 new test lines"]
  J --> L["Enhanced capacity tests"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>c_string.hpp</strong><dd><code>Add string capacity and size method declarations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/c_string.hpp

<ul><li>Add five new method declarations: <code>empty()</code>, <code>utf8_size()</code>, <code>length()</code>, <br><code>max_size()</code>, <code>capacity()</code><br> <li> Include comprehensive documentation for each method<br> <li> Methods follow standard C++ string interface conventions</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/108/files#diff-5e47c64549828a3af3326bf67149f85cf5e336f9492239d74625826ff58bd56a">+78/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>c_string.inl</strong><dd><code>Implement string capacity and size methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

include/core/c_string.inl

<ul><li>Implement <code>empty()</code> method checking first character for null terminator<br> <li> Implement <code>utf8_size()</code> calling external <code>utf8Len()</code> function<br> <li> Implement <code>length()</code>, <code>max_size()</code>, and <code>capacity()</code> methods returning <br><code>size()</code><br> <li> Update assertion conditions to use new <code>empty()</code> method</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/108/files#diff-83492e82f06fe5d99ac148878a31e3bcb9392081ca4179a9595f8f109e5f87fd">+22/-2</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>c_string_test.cpp</strong><dd><code>Add comprehensive test coverage for new methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/core/c_string_test.cpp

<ul><li>Add comprehensive test cases for <code>empty()</code> method (150+ lines)<br> <li> Add extensive test cases for <code>size()</code> method (150+ lines)<br> <li> Add UTF-8 specific tests for <code>utf8_size()</code> method (120+ lines)<br> <li> Add complete test coverage for <code>length()</code>, <code>max_size()</code>, and <code>capacity()</code> <br>methods</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/108/files#diff-61b83cabe56b2b5eca3eb29e11f0be5dd90045ed311cb9f511fc681ccadae5df">+675/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>fix_string_test.cpp</strong><dd><code>Enhance FixedString capacity test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/core/fix_string_test.cpp

<ul><li>Expand <code>max_size()</code> test cases with multiple sections and edge cases<br> <li> Enhance <code>capacity()</code> test cases with comprehensive coverage<br> <li> Add validation for capacity formula and consistency checks<br> <li> Remove outdated refactoring comment</ul>

</details>

  </td>
  <td><a href="https://github.com/ToymanInteractive/toygine2/pull/108/files#diff-cc1ff39c1921067e1cd9ffcf0bc8c8f3d7349f0178f1eb6ddeac099bc0a2d2ae">+297/-10</a></td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: include/core/c_string.hpp [hunk 1]
```diff
@@ -208,8 +208,86 @@ class CString {
...
+    length of zero.
+
+    \return True if the string is empty, false otherwise.
+
+    \note An empty string has size zero.
+    \note An empty string still contains a null terminator.
+    \note This method is equivalent to the expression: ```size() == 0```.
+  */
+  [[nodiscard]] constexpr bool empty() const noexcept;
+
```
[the remark this answers is not in the document]
Comment (dkrutskikh) [reply include/core/c_string.hpp:224]: won't be fixed
